### DISCUSSION TITLE
Running run_sketch.py on Windows

### DISCUSSION THREAD

**Discussion by tabreturn** [py5coding/py5generator#36] [Q&A]
I get an error running run_sketch.py on Windows. I enter something like: 

`py C:\Users\user\Desktop\thonny\env\Lib\site-packages\py5_tools\tools\run_sketch.py C:\Users\user\Desktop\a.py`

and I get --

```
Traceback (most recent call last):
  File "C:\Users\user\Desktop\thonny\env\Lib\site-packages\py5_tools\tools\run_sketch.py", line 41, in <module>
    main()
  File "C:\Users\user\Desktop\thonny\env\Lib\site-packages\py5_tools\tools\run_sketch.py", line 37, in main
    imported.run_code(args.sketch_path, classpath=args.classpath)
  File "C:\Users\user\Desktop\thonny\env\lib\site-packages\py5_tools\imported.py", line 116, in run_code
    _run_static_code(
  File "C:\Users\user\Desktop\thonny\env\lib\site-packages\py5_tools\imported.py", line 140, in _run_static_code
    _run_code(new_sketch_path, classpath, new_process, exit_if_error)
  File "C:\Users\user\Desktop\thonny\env\lib\site-packages\py5_tools\imported.py", line 191, in _run_code
    _run_sketch(sketch_path, classpath, exit_if_error)
  File "C:\Users\user\Desktop\thonny\env\lib\site-packages\py5_tools\imported.py", line 162, in _run_sketch
    sketch_ast = ast.parse(sketch_code, filename=sketch_path, mode='exec')
  File "C:\Users\user\AppData\Local\Programs\Python\Python39\lib\ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "C:\Users\user\AppData\Local\Temp\tmpx8li2fl7\_PY5_STATIC_FRAMEWORK_CODE_.py", line 12
    exec(
         ^
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape
```

Something to do with Windows paths, maybe? Or I'm doing something silly? Any ideas?

**Comment by hx2A** [comment]
@tabreturn no, I think this is a bug, assuming you are using the latest version. See #37 

Thanks for spotting this issue

**Reply by hx2A** [reply]
#37 is fixed now

**Comment by hx2A** [comment]
@tabreturn while doing some testing on my Mac computer I tried using run_sketch and realized that it doesn't and can't work on Macs because py5 can only work on Macs through the Jupyter client with the `%gui osx` magic. It's not possible for me to fix this by inserting `%gui osx` somewhere in run_sketch because that magic works by making adjustments to the IPython kernel and the sketch animation thread. If there's no kernel, there's no kernel for the magic to operate on. I should have realized this before today. I mean, I do explicitly say this in the [Special Notes for Mac Users](https://py5.ixora.io/tutorials/mac-users/), but I didn't realize how that would impact you and what you have done with Thonny. I do regret this.

Currently when I try to use this on my Mac, the terminal just hangs. Should I change it so it throws an exception instead? If it doesn't work, I should at least clearly state that to the user instead of hanging.

Theoretically a fix is possible. There are a few avenues.

* The best choice is to wait for jpype to fix the GUI issues on Mac computers. The problem might be on the py5/Processing side though.
* I can figure out what `%gui osx` does and somehow replicate it outside of the kernel. Previous attempts at this have failed though.
* There's an ugly solution where on Macs I can write code to a temporary notebook file and then launch a process that executes something like:

```
jupyter nbconvert --execute --to notebook --inplace "code.ipynb"
```

That would require Jupyter to be installed along with the py5 kernel. It is achievable though. But not by Sunday. :-( Even if I had 100% of my time available, I still couldn't do that. But I will look into it to see what can be done.

There might be other avenues to pursue. Java on Macs are a huge thorn in my side, but I will stick with it.

**Reply by hx2A** [reply]
Here's what I am going to do. I've pretty much reached the end of what I can accomplish on my own solving the OSX issues. I'm not the right person to solve these kinds of OSX platform specific Java GUI problems. So I am going to create a new repo with some demo code and the documentation and information I've collected. The demo code will clearly outline the problem, what doesn't work, and what should happen instead. Then we can ask people to have a look and contribute ideas, more information, or perhaps a solution. Eventually a solution will be found, and perhaps along the way the problem solving process will contribute to building a community around this project.

**Reply by hx2A** [reply]
@tabreturn I woke up this morning determined to do better. I don't want you to have to say during your presentation that what you are presenting only works on Windows and Linux.

I did some searching and found a new avenue to pursue: [pyobjc](https://github.com/ronaldoussoren/pyobjc). And what do you know, I figured out a new way to run py5 on OSX, without the `%gui osx` magic. I am using that library's [AppHelper](https://pyobjc.readthedocs.io/en/latest/api/module-PyObjCTools.AppHelper.html) tools. And I have a working prototype that can run a Sketch using the default renderer and the OpenGL renderer P2D. And best of all, it isn't a lot of code. This is the solution I've been looking for to solve all of py5's OSX problems.

Incorporating this new idea into all of py5's kernels, modes, IPython magics, etc will take some time, but I can get this new approach working in only run_sketch.py by the weekend. (Currently) I need to make a call to `os._exit(0)` when exiting the Sketch, but run_sketch.py is the one place where that doesn't matter. But I expect to be able to fix that and exit normally. When I can do that, I can apply this approach to all of py5.

One challenge here is the pyobjc library is only installable on OSX machines. So I need to adjust the packaging process so the dependency is only for OSX computers. Or make a different package for OSX.

This is a decent amount of work but I will pull it off by this weekend. Count on it.

**Reply by hx2A** [reply]
@tabreturn I absolutely didn't think I would be saying this 24 hours ago, but I now have a working prototype of run_sketch on OSX.

> I tried using run_sketch and realized that it doesn't and can't work on Macs because py5 can only work on Macs through the Jupyter client with the %gui osx magic.

Very happy to have proven myself wrong.

I didn't have to write a lot of code. It works fine. The new code isn't a hack. And the code changes are minimally invasive, so I am confident I won't break anything for Windows or Linux users.

Tomorrow I will do more testing, and if everything checks out, I'll move forward with the release.

After the release, I will readdress this. What I learned today can be applied elsewhere in py5, but it would be dangerous to attempt to do so right now, because I'll have to make changes to some core parts of the library.

The current state of py5 on OSX is that there are some limitations, but they are documented along with the workarounds. Over the past year OSX functionality has been lagging behind other operating systems but it keeps moving forward. That trend will continue after this release.

**Reply by hx2A** [reply]
> One challenge here is the pyobjc library is only installable on OSX machines. So I need to adjust the packaging process so the dependency is only for OSX computers.

That was easy also.

```
INSTALL_REQUIRES = [
...
    'pyobjc>=7.3;sys_platform=="darwin"',
...
]
```

**Reply by tabreturn** [reply]
Fantastic work @hx2A. Sorry I took a while to respond. I've been slammed thanks to a sudden and strict Covid lockdown in NZ. 

In my presentation, I was just going to mention that macOS support could be expected *sometime* in the future, haha.

I'll do some testing -- unfortunately, I cannot get back into my office right now, so I do not have Mac access (only Win and Linux).

**Reply by villares** [reply]
Today I tried a clean install on another Windows machine (using Thonny 3.3.7-alt(Python 3.9). 

I'm having trouble with the py5 installation as it tries to build `noise`:
```
Running setup.py clean for noise
Failed to build noise
Installing collected packages: noise, line-profiler, jpype1, ipywidgets, py5
    Running setup.py install for noise ... error
    ERROR: Command errored out with exit status 1:
```

**Reply by tabreturn** [reply]
Hmm. Looks like the `noise` package hasn't been updated since 2015 -- https://pypi.org/project/noise/#history

**Reply by hx2A** [reply]
> Fantastic work @hx2A. Sorry I took a while to respond. I've been slammed thanks to a sudden and strict Covid lockdown in NZ.

Hang in there. I do think your country is doing a good job of managing the pandemic. Certainly better than what is happening over here.

> I'll do some testing -- unfortunately, I cannot get back into my office right now, so I do not have Mac access (only Win and Linux).

It happens I didn't have access to a Mac to use for this project until March of this year. I wasn't able to borrow one because of the pandemic, so eventually I decided to just buy a used one for testing purposes. Since I haven't been using py5 on OSX for as long as Windows and Linux, the features are behind the other platforms. I do want to get py5's functionality equal across the 3 platforms.

I thought today would be some easy testing and do the release, but of course everything went wrong. The code I wrote yesterday wasn't working consistently for reasons that seemed to be related to the environment and not the code. After much flailing about, I figured out how to get it work reliably.

I'll do some more testing tomorrow and will aim to do the release on Friday.

Eventually I'd like to have a stress-free release. This was not that release.

> In my presentation, I was just going to mention that macOS support could be expected _sometime_ in the future, haha.

Globally Macs have small market share, but in the US, Macs are very popular with the creative tech crowd. When I was at NYU, there were only a handful of students who did not use a Mac. I do know how important Macs are to people here, and I wouldn't want people's initial opinion of py5 to be clouded by OSX issues.

**Reply by hx2A** [reply]
> Today I tried a clean install on another Windows machine (using Thonny 3.3.7-alt(Python 3.9).
> 
> I'm having trouble with the py5 installation as it tries to build `noise`:
> 
> ```
> Running setup.py clean for noise
> Failed to build noise
> Installing collected packages: noise, line-profiler, jpype1, ipywidgets, py5
>     Running setup.py install for noise ... error
>     ERROR: Command errored out with exit status 1:
> ```

The noise library uses c code in addition to Python code. Therefore, if the wheel has not been compiled for a platform, pip must download the source and comple it itself. I see here:

https://pypi.org/project/noise/#files

That the wheels stop at python 3.4. That's unfortunate. I wish the maintainer would at least add new wheel distributions.

@villares how are you installing this? If you install it through Anaconda, it can obtain a compiled wheel from conda-forge. No compilation necessary. If you are installing through pip, it will need to compile the source first. In the error messages somewhere it should provide a link to somewhere at Microsoft that provides links for a compiler. It is a bit of a drag to install those things, but they work. Do you see any messages like that?

**Reply by hx2A** [reply]
> Hmm. Looks like the `noise` package hasn't been updated since 2015 -- https://pypi.org/project/noise/#history

How much do the Perlin or Simplex noise algorithms change from year to year? Does the code need to be updated?

The noise library isn't a lot of code, so perhaps there has not been a need to change anything.

I do wish they'd continue doing releases though. It shouldn't be a lot of work for stable code base.

**Reply by tabreturn** [reply]
> How much do the Perlin or Simplex noise algorithms change from year to year? Does the code need to be updated?

True. But if they did release another version, I guess wheels would be nice.

> I do know how important Macs are to people here, and I wouldn't want people's initial opinion of py5 to be clouded by OSX issues.

Agreed. Also, entire, Mac-only labs are pretty commonplace in design/art/creative schools the world over. Like Processing, I'm sure many of the people who'd find py5 appealing are educators (and by extension, students) in those spaces.

**Reply by villares** [reply]
> @villares how are you installing this? If you install it through Anaconda, it can obtain a compiled wheel from conda-forge. No compilation necessary. If you are installing through pip, it will need to compile the source first. In the error messages somewhere it should provide a link to somewhere at Microsoft that provides links for a compiler. It is a bit of a drag to install those things, but they work. Do you see any messages like that?

I see, I was trying emulate the kind of "minimal" install I would have on classroom  machines, just Thonny with it's own Python, so just `pip` and no `conda` at this point.

Could you "vendorize" noise, include your own wheels? Could you use Processing's noise for now (I suppose you want to avoid this for performance issues?), just to make things simpler?

**Reply by hx2A** [reply]
@villares What version of Python are you using? Is your error related to [this noise issue](https://github.com/caseman/noise/issues/32)?

I'm concerned that the library maintainer is not responding to that project's issues. Perhaps I made a bad choice by selecting this library to be included in py5.

> Could you "vendorize" noise, include your own wheels? Could you use Processing's noise for now (I suppose you want to avoid this for performance issues?), just to make things simpler?

I'm going to have to do something. Are any of the project's forks being better maintained?

This isn't a change I can address for the next release (due tomorrow-ish) but I will look into what options I have for after.

**Reply by hx2A** [reply]
> > I do know how important Macs are to people here, and I wouldn't want people's initial opinion of py5 to be clouded by OSX issues.
> 
> Agreed. Also, entire, Mac-only labs are pretty commonplace in design/art/creative schools the world over. Like Processing, I'm sure many of the people who'd find py5 appealing are educators (and by extension, students) in those spaces.

I'm happy to be using Linux for my creative projects. There are pro's and con's to being on a different platform as everyone else, but for me it is a great choice. I do believe that if I was a Mac person, py5 would not exist. OSX would have presented too many hurdles early on in my initial project experiments that would have either led me in a different direction or caused me to think it was not possible.

**Reply by hx2A** [reply]
> This isn't a change I can address for the next release (due tomorrow-ish) but I will look into what options I have for after.

I created a new issue #38 to track this.

**Reply by villares** [reply]
> @villares What version of Python are you using? Is your error related to [this noise issue](https://github.com/caseman/noise/issues/32)?

I was using Python 3.9, I think it could be the same issue. I won't have access to Windows today to probe more.

Just out of curiosity, what is the motive for not using the noise function provided by Processing?
Is it outside the core machinery you are accessing, willing to access?

**Reply by hx2A** [reply]
@villares Processing's noise functionality works great but py5 does not use it strictly for performance reasons. You'll also notice it does not use Processing for the other random functions or the math related functions like `sin` or `atan2`. All of that is done in numpy. The performance difference is significant, particularly if you are using these things in a loop, which typically one would do when using them.

**Reply by hx2A** [reply]
@villares py5 also depends on the line_profiler package, which also requires a c compiler. That library is essential and is not replaceable.

Conveniently, as of the latest release of line_profiler, compiled versions for all platforms and python versions are uploaded to pypi. Before 3.3, they didn't do that. I hope they continue doing that.

It seems that if I get rid of the noise library dependency, I can simplify the py5 installation process. It is the only one that requires compilation. I won't do that for this week's release but I will investigate what can be done for the release after this one.

**Reply by jeremydouglass** [reply]
Since it is specifically a performance dependency, it might be nice to make it optional -- a common pattern for performant requirements (which are also often C). Try, then fall back to Processing noise. That way people who care about the performance can get it working, but people who don't will still be able to run py5.

An example of this design pattern is the natsort library, which is designed to be highly optimized with fastnumbers, but still runs in the worst case.

https://github.com/SethMMorton/natsort#optional-dependencies

**Reply by hx2A** [reply]
@jeremydouglass yes, this can be an effective design pattern. Easy to install and get working for everyone, while additional features and performance available for those who install optional requirements.

There is an extra complication that needs to be addressed for that design idea to work here: the two different noise algorithms / libraries need to generate exactly the same noise values for the same inputs. If I create a sketch on my computer and fiddle with the noise parameters to get a certain effect on my computer, I want you to see the exact same thing on your computer, regardless of what your setup is.

I do know that Processing's noise function and the noise library we are discussing do not generate identical numbers. The vnoise library suggested by @villares says it is a pure Python port of the noise library, so perhaps those two generate identical numbers. This would need to be tested first though. However, I'm skeptical that including the noise library at all is a good choice. In addition to the lack of current Python wheels, the library has some open issues that include seg faults, and I want py5 to stay away that.

More testing and examination of the various libraries needs to be done to learn more and then find the best path forward.

**Reply by jeremydouglass** [reply]
True. By coincidence, I became interested in generating a (very
non-performant) cross-mode library a few years ago that would provide
reproducible random number seed sequences for Processing, Processing.py,
p5.js, and Processing.R called RandAnywhere / CommonRandom.

https://discourse.processing.org/t/cross-language-random-numbers/3474

If I could have dropped in a vended library set that did this I would have,
but I didn't find one, so I hand implemented XORshift in each language for
the prototype. Of course, people trying to move code between Processing.py
or pyp5 and the py5 (or Java or p5.js and py5) are still in for PRNG
disappointment. On the other hand, if your goal was for a drop-in to match
the Java PRNG exactly, it is not overly complex and extremely well
documented.

**Reply by hx2A** [reply]
@jeremydouglass I like your random number library. It's kind of like a Rosetta Stone for random numbers. There's a lot for people to learn here, both from the technicality that goes into actually generating random numbers and your attention to detail.

**Reply by hx2A** [reply]
> I'll do some more testing tomorrow and will aim to do the release on Friday.
> 
> Eventually I'd like to have a stress-free release. This was not that release.

@tabreturn I tested the new release on Linux and OSX and everything works OK, including run_sketch.py. Although nothing changed for Windows users since when I tested it on Windows earlier this week, I'll test it again anyway tomorrow morning just to be sure, and will do the release after that.

I'm going to be very happy when I get this release out the door.

**Reply by tabreturn** [reply]
Fantastic work, @hx2A. I've tested on Linux (Mint 20.2) and Windows 10 without issue.

**Reply by hx2A** [reply]
@tabreturn The new release is live. The run_sketch utility should work correctly on all platforms.

While testing I noticed that py5bot has some odd issues on Windows when using an OpenGL renderer. The window doesn't seem to close when it is supposed to and there are odd exceptions in the Jupyter logs. I'm not sure if it is my computer or a real issue. Can you give it a try and open an issue if you experience it also?

**Reply by tabreturn** [reply]
Using Thonny+py5, static-mode P2D/P3D sketches work fine using the run_sketch.py on Linux. But on Windows, one has to use a draw() function (otherwise, shapes/forms disappear instantly, seemingly drawn over by the background fill).

Regarding Jupyter: I wasn't aware that it would open a window for OpenGL sketches. I tried a sketch with a `size(200, 200, P3D)` using the py5bot kernel, which renders fine (as an embedded notebook graphic). If you'd like me to try something else, let me know what steps I should carry out.

**Reply by hx2A** [reply]
> But on Windows, one has to use a draw() function (otherwise, shapes/forms disappear instantly, seemingly drawn over by the background fill).

Hmmm, this appears to be a bug. On Windows, when using the P2D renderer, some draw commands don't work when they are in setup. So for example:

```
def setup():
    size(200, 200, P2D)
    background(0)
    fill(255)
    rect(10, 20, 130, 140) # this rectangle appears and gets overwritten

def draw():
    rect(mouse_x, mouse_y, 10, 10)
```

This doesn't happen on Linux or OSX or when using the default renderer on Windows. This is not limited to run_sketch either; it happens in the regular notebook kernel also.

This seemed too strange and I don't have any immediate ideas why this bug would happen like this only on Windows. I tested the analogous code on the latest Processing 4 release and it works correctly, so the problem is in py5 somewhere.

Can you experiment with this a bit more and open an issue? I will need to think about this some more to come up with ideas for how to investigate this.

While testing run_sketch I had been testing Sketches in functionless "static mode" or Sketches with both a setup and draw function. I tested with only a setup function and I see that run_sketch has an issue on OSX. When using the default renderer, if the setup function is defined and there is no draw function, the sketch window will appear but then close on its own after a few seconds. A few days ago I did discover this would happen for static mode sketches that used the default renderer, and I fixed it, but only for static mode sketches, not sketches with only a setup function. Fixing this would require me to fiddle with my newly written pyobjc code, which I will need to take my time with. It doesn't seem that critical though, so let's leave it be for now and I'll fix it in a future release.

**Reply by hx2A** [reply]
@tabreturn I now know the cause of this Windows issue (see #5) and I know how to fix it, but not in a way that does not break something else. But fixing this is achievable, and I'll work on it for the next release.

**Comment by villares** [comment]
Cool! No hurry!

Someone else had issues with the `noise` library and created [vnoise](https://github.com/plottertools/vnoise) 
Might be worth a look! It could be too slow for our use case... it gets fast only when used in a "vectorized" numpyish call :(

**Reply by hx2A** [reply]
This is a great suggestion. Thank you for the research!

The current py5 noise functionality supports vectorized noise generation through use of `np.vectorize`. And generally I think vectorized operations are the way to go for py5, and Python in general. This replacement being slower only matters if someone uses it in a loop. If someone needs many noise values and cares about performance, they can use the vectorized approach instead.

From looking at that the vnoise author's twitter feed it seems they like using Python and Processing, so perhaps they would also be interested in using py5, and would be happy that we are using their library.

**Reply by villares** [reply]
WHOA! That's a great twitter catch you made, huh? Followed them!

**Reply by hx2A** [reply]
> WHOA! That's a great twitter catch you made, huh? Followed them!

I'm slowly catching on to the twitter world...after this next release, I'll start using the py5coding account.
